### PR TITLE
openmoji-black,openmoji-color: init at 12.0.0

### PR DESCRIPTION
**Pull request by fgaz**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/data/fonts/openmoji/default.nix
Comment (fgaz) [inline pkgs/data/fonts/openmoji/default.nix]: woops, forgot about the metas. pushing a fix shortly.

Comment (fgaz) [reply pkgs/data/fonts/openmoji/default.nix]: Done

File: pkgs/data/fonts/openmoji/default.nix
Comment (dtzWill) [inline pkgs/data/fonts/openmoji/default.nix]: This bash reimplementation is broken, unfortunately, but happily it's an easy fix for the issue I noticed.  Here's the resulting commit: 0a5b480de54b71f95eefb77fbbaf795a6734df53 (on my "fork").

This may not be enough, as the resulting color variant isn't color and I'm not sure all the expected glyphs end up included.  But I can imagine reasons my tree is to blame-- do the resulting files work for you? What have you tried them with?
(grabbing the font files provided for the release don't seem to work for me either, which also suggests it's something on my end...)

I had a fear that hexcodes actually needed to include the range of values between the hyphenated pair but offhand I'm not sure whether that's what the javascript this is based on actually does O:).

----

Perhaps we should simply fetch and use the provided generated font files? (assuming they work for folks?)

**Review by dtzWill — CHANGES_REQUESTED** [review]
Bash reimplementation needs some attention, resulting files don't work for me but that may be an issue locally.  See my comment for details.

File: pkgs/data/fonts/openmoji/default.nix
Comment (fgaz) [reply pkgs/data/fonts/openmoji/default.nix]: Oops, yeah, thanks for the fix, I'll cherry pick your commit shortly.

The weird thing is that both your version and mine do work for me. Probably because that part only handles missing glyphs.

I tested it on Firefox, one of the few programs that support color fonts. Install the font then go to about:config and set `font.name-list.emoji` to `OpenMoji`. Here's the result:

2019-10-02-152635_957x419_scrot

A bit glitchy on some glyphs, but it works!

---

I think the hypenated pairs mean that the emoji is for a ligature between the two glyphs. For example 🇮 (REGIONAL INDICATOR SYMBOL LETTER I) + 🇹 (REGIONAL INDICATOR SYMBOL LETTER T) = 🇮🇹 (italian flag) (try to copy this text and delete the space between the two letters: 🇮 🇹)

Comment (fgaz) [reply pkgs/data/fonts/openmoji/default.nix]: Cherry-picked

**Comment by fgaz** [comment]
Ping 1F60A

**Comment by fgaz** [comment]
Updated to 12.1.0

**Comment by fgaz** [comment]
Not stale, stale-bot. Unfortunately though 12.3.0 segfaults scfbuild, so I cannot update this pr to that version.

**Comment by fgaz** [comment]
Ugggh, 12.4.0 is segfaulting scfbuild too, maybe I should just fetch the prebuilt font

**Comment by OPNA2608** [comment]
I would *love* to finally have these fonts in stable, can you check the state of this again please? I checked out this PR locally, bumped `scfbuild` to 2.0.0 (https://github.com/13rac1/scfbuild/commit/6d84339512a892972185d894704efa67dd82e87a) & dropped the patch in the `postPatch` phase and FWIW both the black and color builds succeed.

**Comment by fgaz** [comment]
I just tried but it doesn't seem to build for me. Maybe I missed something though. Let's see what ofborg says

**Comment by OPNA2608** [comment]
I bumped those versions without rebasing before, but now I see the crash you were mentioning. I've debugged this down to a problem in `fontforge`:

> ```
> malloc_consolidate(): invalid chunk size
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x00007ffff6af63c0 in SplineSetStroke () from /nix/store/1k5kvnxjfrvpqgwbadmisjpygnbbamfh-fontforge-20200314/lib/libfontforge.so.4
> ```

Upstream issue https://github.com/fontforge/fontforge/issues/4315 seems related, should we chime in there with this "real world" example of that crash?

File: pkgs/data/fonts/openmoji/default.nix
Comment (JustinLovinger) [inline pkgs/data/fonts/openmoji/default.nix:30]: Upstream no longer has helpers/export-svg-font.js <https://github.com/hfg-gmuend/openmoji/tree/12.4.0/helpers>. Also, if possible, I recommend patching source instead of copying-pasting script contents. That is less brittle.

It looks like we are now supposed to run `npm run generate-font`: <https://github.com/hfg-gmuend/openmoji/tree/12.4.0/helpers#big-helpers>.

File: pkgs/data/fonts/openmoji/default.nix
Comment (JustinLovinger) [inline pkgs/data/fonts/openmoji/default.nix:4]: This argument is not used.

**Review by justinlovinger — CHANGES_REQUESTED** [review]
I think the build should be refactored to use the upstream `npm run generate-font`. It looks like we are currently building in an unsupported way. That is probably why the build fails for us, but presumably succeeds upstream.

File: pkgs/data/fonts/openmoji/default.nix
Comment (fgaz) [reply pkgs/data/fonts/openmoji/default.nix:30]: I didn't copypaste, I reimplemented it entirely. The reason why I did that instead of using their scripts is that then we'd have to first package the node stuff too (and I see there's some recursive node-bash-_docker_ calls :-/ ), which is a pain.

**Comment by fgaz** [comment]
> admit defeat and use the prebuilt TTFs

:white_flag: :white_flag: :white_flag:
I'll switch this to the prebuilts soon

File: pkgs/data/fonts/openmoji/default.nix
Comment (OPNA2608) [reply pkgs/data/fonts/openmoji/default.nix:30]: > Upstream no longer has helpers/export-svg-font.js

It was [moved](https://github.com/hfg-gmuend/openmoji/commit/3811fc0a9518db8f53453a8e7f341b7bd29dd152) to [`helpers/generate-font-glyphs.js`](https://github.com/hfg-gmuend/openmoji/blob/3811fc0a9518db8f53453a8e7f341b7bd29dd152/helpers/generate-font-glyphs.js).

Comment (JustinLovinger) [reply pkgs/data/fonts/openmoji/default.nix:30]: @fgaz I believe we can install node dependencies with `node2nix` and patch out the `docker` calls, either with `sed` or `patches`.

**Comment by OPNA2608** [comment]
> That is probably why the build fails for us, but presumably succeeds upstream.

Afaict upstream builds with a Docker image based on Ubuntu 18.04 (fontforge 2017*something*) while we have the latest fontforge release packaged instead (fontforge 202003*something*). With an older nixpkgs revision I was able to build this package without problems, since \~ the bump of fontforge it started crashing as I coarsely bisected earlier.

I wasn't sure where exactly to report this back then and was abit stressed with tasks. It's obviously a crash within fontforge as noted earlier, so I'll visit the linked issue later and add our crash example, [assuming this is the same issue](https://github.com/fontforge/fontforge/issues/4315). Though their report implies that it's due to bad input and will cause visual problems even if it's fixed, so informing the openmoji team of this problem would prolly be a good idea too.

> Given that the input is strange and could/should have been a closed contour, I think the fix can focus on avoiding the crash and need not return an ideal result.

I have a hunch they'll tell us to use their handy docker file or prebuilt font files though :stuck_out_tongue:.

**Comment by JustinLovinger** [comment]
@fgaz it looks like both Noto and Twemoji use prebuilt TTFs. I guess there is precident. It would be nice to build from source, but I'm not sure what the practical benefits are.

**Comment by fgaz** [comment]
@JustinLovinger one of the twemojis does, but the other is built from source. Unfortunately I'll have to use the prebuilts in that one too because of #97871

**Comment by JustinLovinger** [comment]
I'm not sure what the current plan for this is, but the fontforge issue appears fixed: <https://github.com/fontforge/fontforge/issues/4315>.

**Comment by OPNA2608** [comment]
The latest release of openmoji builds fine (haven't tried using the font) if you first pull in this patch for fixing the fontforge crash (and deal with the rebuilds this causes): https://github.com/fontforge/fontforge/commit/69e263b2aff29ad22f97f13935cfa97a1eabf207

Do we still want to bother with a source build @fgaz? If yes then I'd like to see the reimplemented script abandoned in favour of getting upstream's version working like the above comments mentioned, if possible. I believe at some point the build date gets inserted into the font too so this whole process doesn't even give reproducible results?

I'd be fine with repackaging their prebuilt versions to save everyone some headache & finally get this merged, it's just a font after all.

**Comment by kevincox** [comment]
I tried applying the fontforge patch mentioned and this package built perfectly! https://github.com/NixOS/nixpkgs/compare/master...kevincox:openmoji (feel free to take the commits I added and cherry-pick them here).

It also didn't take much resources to build so I would love it if we can keep building from source 🥺

**Comment by fgaz** [comment]
Ugh of course it's a mass rebuild... I guess I'll split the scfbuild update into a separate pr.

> I'd like to see the reimplemented script abandoned in favour of getting upstream's version working

If someone else does and documents it, I'll try to maintain it, but my preference is to stay clear of node stuff.

Thanks @OPNA2608 and @kevincox for the contributions!

File: pkgs/tools/misc/fontforge/default.nix
Comment (erictapen) [inline pkgs/tools/misc/fontforge/default.nix:35]: :+1: for the patch.

**Comment by fgaz** [comment]
> Only two NPM dependencies

_direct_ dependencies :)

**Comment by OPNA2608** [comment]
Those two npm packages, once added according to https://nixos.org/manual/nixpkgs/stable/#node.js, are enough to get everything running with their JS script.

```nix
  nativeBuildInputs = [
    scfbuild
    nodejs
    nodePackages.glob
    nodePackages.lodash
  ];
  
  buildPhase = ''
    runHook preBuild
    
    node helpers/generate-font-glyphs.js
    
    cd font
    scfbuild -c scfbuild-${variant}.yml
    
    runHook postBuild
  '';
```

File: pkgs/tools/misc/fontforge/default.nix
Comment (SuperSandro2000) [reply pkgs/tools/misc/fontforge/default.nix:35]: If this is causing the mass rebuild than it would be good we can do a PR that targets staging with it and wait until it is in master.

File: pkgs/tools/misc/scfbuild/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/misc/scfbuild/default.nix:19]: Please delete the patch.

Comment (fgaz) [reply pkgs/tools/misc/scfbuild/default.nix:19]: Done! (`fixup! scfbuild: 1.0.3 -> 2.0.0`)

File: pkgs/tools/misc/fontforge/default.nix
Comment (kevincox) [reply pkgs/tools/misc/fontforge/default.nix:35]: This patch has been upstreamed. We can drop this commit from the PR.

**Review by kevincox — APPROVED** [review]

**Comment by fgaz** [comment]
Oh nooo a conflict... it took literally multiple hours to regenerate node-packages.nix D:

**Comment by fgaz** [comment]
Oh, and i guess we should wait for @kevincox's patch to land in master before merging this

**Comment by kevincox** [comment]
Alternatively I think it is fine to target staging.

**Comment by kevincox** [comment]
@fgaz do you mind if I merge this into staging?

**Comment by kevincox** [comment]
Ah, and it appears someone fixed the merge-conflict on staging so I don't need to regen that  🎉

**Comment by fgaz** [comment]
@kevincox ok. could you preserve the intermediate commit with the bash implementation?

**Comment by kevincox** [comment]
I was just going to merge the commits as is. (And re-gen the node-packages)

**Pull request merged**
